Examples: Toolbar with Menu

lluia opened the pull request:
## Summary

![toolbar](https://user-images.githubusercontent.com/5938217/82118261-4eab3680-9776-11ea-92ea-6834a0f2218d.gif)


As part of the documentation effort ono #626 this PR adds examples for a [Toolbar with Menu](https://reakit.io/docs/toolbar/#toolbar-with-menu).

I tried to stay consistent with existing examples 💅🏻

## Notes

I can remove the **emojis from the example** if needed, I think they add a bit of fun to it, however, I tend to over-use them 😄

lluia commented:
I tried to structure the tests on **smaller units** ( _for better readability and debugging_ ), however, I encountered the following issue:

<img width="766" alt="Screenshot 2020-05-16 at 12 51 46" src="https://user-images.githubusercontent.com/5938217/82118295-8c0fc400-9776-11ea-807e-c1690d97051f.png">

When running the test on isolation it passed 🤔 . Hence I refactored them into a single one. I didn't dig deeper maybe @diegohaz you're familiar with this already?

diegohaz commented:
It seems that it's still throwing those errors. I'd try to use some `await wait()` to check if the menu is visible and the initial focus has been given to the menu item. You can use this as a reference: https://github.com/reakit/reakit/blob/075b919f084b4dbf0d5dd1fe8e12598bd48fad86/packages/reakit/src/Menu/__examples__/MenuWithSubmenu/__tests__/index-test.tsx#L5-L110

diegohaz commented:
And I agree with you that it looks cooler, but for the sake of code simplicity, I'd remove the emojis 😆

lluia commented:
@diegohaz I re-structured the tests and it seems finnally 🟢 ( no need for `wait()` ) , I removed emojis as well 👍🏻

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/index.tsx:
```diff
@@ -0,0 +1,48 @@
+import { useMenuState, MenuButton, Menu, MenuItem } from "reakit/Menu";
+import { Button } from "reakit/Button";
+
+const MoreItems = React.forwardRef((props: ToolbarProps, ref) => {
```
Didn't test it, but I think you don't need to explicitly cast `ref` by doing this:
```suggestion
const MoreItems = React.forwardRef<HTMLButtonElement, ToolbarProps>((props, ref) => {

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/index.tsx:
```diff
@@ -0,0 +1,48 @@
+        {...menu}
+        {...props}
+        ref={buttonRef}
+        as={Button}
```
`MenuButton` already uses `Button`
```suggestion
```

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/__tests__/index-test.tsx:
```diff
@@ -0,0 +1,78 @@
+import { render, press, click } from "reakit-test-utils";
+import ToolbarWithMenu from "..";
+
+describe("<ToolbarWithMenu />", () => {
```
Maybe we should have a lint rule for that, but, except for very specific cases, I try not to nest test suites with `describe` to make things simpler as the file itself already provides the context.
```suggestion
```

lluia replied:
Ah I forgot 😓 ,  that's my habit of sticking to this style, let me refactor them to adhere to rest of the codebase 🙏🏻

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/__tests__/index-test.tsx:
```diff
@@ -0,0 +1,78 @@
+    expect(text("Lemons")).not.toBeVisible();
+  });
+
+  it("can navigate toolbar items through keyboard", () => {
```
I tend to use always `test`. I think we can enforce this with [this eslint plugin](https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/consistent-test-it.md), but this can be done in another PR. :)
```suggestion
  test("can navigate toolbar items through keyboard", () => {
```

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/__tests__/index-test.tsx:
```diff
@@ -0,0 +1,78 @@
+    expect(text("Pears")).toHaveFocus();
+  });
+
+  it("can open and close the menu through mouse", () => {
```
```suggestion
  test("can open and close the menu through mouse", () => {
```

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/__tests__/index-test.tsx:
```diff
@@ -0,0 +1,78 @@
+    expect(text("Pears")).not.toBeVisible();
+  });
+
+  it("can open menu, navigate it and close it through keyboard", () => {
```
```suggestion
  test("can open menu, navigate it and close it through keyboard", () => {
```

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/__tests__/index-test.tsx:
```diff
@@ -0,0 +1,78 @@
+    press.ArrowRight();
+    expect(text("Other Fruits")).toHaveFocus();
+
+    press.Enter();
```
Nice! 👍
I guess only `press.ArrowDown()` and `press.ArrowUp()` would need `wait()`.

lluia commented:
@diegohaz addressed your suggestions 💯 on my last commit 💭 , notice I used [`findByText`](https://testing-library.com/docs/dom-testing-library/api-queries#findby) rather than [`wait`](https://testing-library.com/docs/dom-testing-library/api-async#wait-deprecated-use-waitfor-instead) as it feels more idiomatic for **RTL** 💅🏻

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/__tests__/index-test.tsx:
```diff
@@ -0,0 +1,114 @@
+
+  press.ArrowUp();
+
+  const firstMenuItem = await findText("Lemons");
```
Nice! I didn’t know about that! 👍

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/__tests__/index-test.tsx:
```diff
@@ -0,0 +1,114 @@
+  expect(text("Lemons")).not.toBeVisible();
+});
+
+it("can navigate toolbar items through keyboard", () => {
```
One remaining guy
```suggestion
test("can navigate toolbar items through keyboard", () => {
```

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/index.tsx:
```diff
@@ -0,0 +1,45 @@
+const MoreItems = React.forwardRef<HTMLButtonElement, ToolbarProps>(
+  (props, ref) => {
+    const menu = useMenuState({ placement: "bottom-end" });
+    const buttonRef = ref as RefObject<HTMLButtonElement>;
```
Now that the ref is properly typed I think this can be removed

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/index.tsx:
```diff
@@ -0,0 +1,45 @@
+import React, { RefObject } from "react";
```
```suggestion
import * as React from "react";
```

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/index.tsx:
```diff
@@ -0,0 +1,45 @@
+import React, { RefObject } from "react";
+import { useToolbarState, Toolbar, ToolbarItem } from "reakit/Toolbar";
+import { useMenuState, MenuButton, Menu, MenuItem } from "reakit/Menu";
+import { Button } from "reakit/Button";
+import { ToolbarProps } from "../../Toolbar";
```
```suggestion
import {
  useToolbarState,
  Toolbar,
  ToolbarItem,
  ToolbarProps,
} from "reakit/Toolbar";
import { useMenuState, MenuButton, Menu, MenuItem } from "reakit/Menu";
import { Button } from "reakit/Button";
```

diegohaz commented on packages/reakit/src/Toolbar/__examples__/ToolbarWithMenu/index.tsx:
```diff
@@ -0,0 +1,45 @@
+const MoreItems = React.forwardRef<HTMLButtonElement, ToolbarProps>(
+  (props, ref) => {
+    const menu = useMenuState({ placement: "bottom-end" });
+    const buttonRef = ref;
+
+    return (
+      <>
+        <MenuButton
+          {...menu}
+          {...props}
+          ref={buttonRef}
```
We can just pass the ref directly.
```suggestion
    return (
      <>
        <MenuButton
          {...menu}
          {...props}
          ref={ref}
```

lluia commented:
@diegohaz addressed your feedback 💯 ,  I think as you said now should be good to go 👍🏻
